### PR TITLE
Sema: Fix a couple of issues related to variance of protocol 'Self'

### PR DESCRIPTION
**Pull request by slavapestov**
The first one is a fix for a regression from https://github.com/apple/swift/pull/41677. The second one is rdar://problem/74944514, which is a follow-up to https://github.com/apple/swift/pull/41545.

I'm wondering if the new variance check in the conformance checker, findExistentialSelfReferences() and Sema's doesMemberHaveUnfulfillableConstraintsWithExistentialBase() can all be unified somehow.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/decl/protocol/existential_member_accesses_self_assoctype.swift [hunk 1]
```diff
@@ -579,6 +581,19 @@ do {
...
+}
+
+// Make sure this also works in a generic context!
+struct G<X, Y, Z> {
+  func doIt() {
+    let exist: any UnfulfillableGenericRequirements
+
+    exist.method8(false)
+    // expected-error@-1 {{member 'method8' cannot be used on value of type 'any UnfulfillableGenericRequirements'; consider using a generic constraint instead}}
+  }
```
Comment (AnthonyLatsis) [inline test/decl/protocol/existential_member_accesses_self_assoctype.swift:596]: Oh, if only these declarations could just be anonymous.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Review by AnthonyLatsis — APPROVED** [review]
> I'm wondering if the new variance check in the conformance checker, findExistentialSelfReferences() and Sema's doesMemberHaveUnfulfillableConstraintsWithExistentialBase() can all be unified somehow.

Unifying the first and last sounds simple enough, I could see to that in a follow-up. I think the type signature check should have its own entry point: apparently there are places like this one where only generic requirements are of interest ([and vice versa?](https://github.com/apple/swift/blob/512ebc52a64c844589e4b42dbd5ea36da655fa28/lib/Sema/TypeCheckProtocol.cpp#L1045)).

**Pull request merged**